iceoryx2-bb/posix/src/periodic_timer.rs

```diff
@@ -0,0 +1,235 @@
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PeriodicTimerBuilder {
+    /// Creates a new builder.
+    pub fn new() -> Self {
+        Self {
+            clock_type: ClockType::default(),
+        }
+    }
+
+    /// Defines the [`ClockType`] that is used for time measurements. By default it is
+    /// [`ClockType::default()`].
+    pub fn clock_type(mut self, value: ClockType) -> Self {
+        self.clock_type = value;
+        self
+    }
+
+    /// Creates a new [`PeriodicTimer`]
+    pub fn create(self) -> Result<PeriodicTimer, TimeError> {
+        let start_time = fail!(from "PeriodicTimer::new()", when Time::now_with_clock(self.clock_type),
+                                "Failed to create PeriodicTimer since the current time could not be acquired.");
+        let start_time = start_time.as_duration().as_nanos();
+
+        Ok(PeriodicTimer {
+            timeouts: RefCell::new(vec![]),
+            id_count: IoxAtomicU64::new(0),
+            clock_type: self.clock_type,
+            last_iteration: RefCell::new(start_time),
+        })
+    }
+}
+
+#[derive(Debug)]
+struct Attachment {
+    index: u64,
+    period: u128,
+    start_time: u128,
+}
+
+impl Attachment {
+    fn new(index: u64, period: u128, clock_type: ClockType) -> Result<Self, TimeError> {
+        let start_time = fail!(from "Attachment::new()", when Time::now_with_clock(clock_type),
+                                "Failed to create PeriodicTimer attachment since the current time could not be acquired.");
+        let start_time = start_time.as_duration().as_nanos();
+
+        Ok(Self {
+            index,
+            period,
+            start_time,
+        })
+    }
+
+    fn reset(&mut self, clock_type: ClockType) -> Result<(), TimeError> {
+        let start_time = fail!(from "Attachment::new()", when Time::now_with_clock(clock_type),
+                                "Failed to reset PeriodicTimer attachment since the current time could not be acquired.");
+        self.start_time = start_time.as_duration().as_nanos();
+        Ok(())
+    }
+}
+
+/// The [`PeriodicTimer`] allows the user to attach multiple periodic timers with
+/// [`PeriodicTimer::add()`], to wait on them by acquiring the waiting time to the next timer
+/// with [`PeriodicTimer::next_iteration()`] and to acquire all missed timers via
+/// [`PeriodicTimer::missed_timers()`].
+#[derive(Debug)]
+pub struct PeriodicTimer {
+    timeouts: RefCell<Vec<Attachment>>,
+    id_count: IoxAtomicU64,
+    last_iteration: RefCell<u128>,
+
+    clock_type: ClockType,
+}
+
+impl PeriodicTimer {
+    /// Adds a cyclic timeout to the [`PeriodicTimer`] and returns an [`PeriodicTimerGuard`] to
+    /// identify the attachment uniquely.
+    /// [`PeriodicTimer::next_iteration()`] will schedule the timings in a way that the attached
+    /// timeout is considered cyclicly.
+    pub fn cyclic(&self, timeout: Duration) -> Result<PeriodicTimerGuard, TimeError> {
```

Comment:
```suggestion
    pub fn schedule(&self, timeout: Duration) -> Result<PeriodicTimerGuard, TimeError> {
```

A `PeriodicTimer` with function `cyclic` is confusing - the difference between the two is not obvious.

What do you think of the name `schedule` for the function? i.e. it schedules a new interval.

The `PeriodicTimer` returning a `PeriodicTimerGuard` is also unintuitive to me. With `PeriodicTimerGuard`, it sounds like it is related to the lifetime of the `PeriodicTimer`, not the specific `Attachments` or `Intervals` that it are triggered.

Reply:
@elfenpiff How is that different from it being a `PeriodicTimer` ?

Reply:
The `PeriodicTimerGuard` is a RAII guard so that the user does not forget to remove a cyclic timeout when it is no longer required - otherwise one would leak those into the `PeriodicTimer`.

The `Attachment` is an internal struct that the user never encounters.

Reply:
@elfenpiff Maybe replied to the wrong comment?

Reply:
Resolved by renaming `PeriodicTimer` to `Timer`.